### PR TITLE
Feature/8 extend uipageviewcontroller protocols

### PR DESCRIPTION
**Pull request by barthopster**
***Context***
By extending the `Segnify` and `UIPageViewController` protocols, we create the possibility to be informed about several events, like `Segment` selections and `UIPageViewController` swipes.

We could've used the existing `Segnify` and `UIPageViewController` protocols by overriding the current delegate implementations in `Segnify`, but we would introduce the risk of breaking the core functionality of `Segnify`. The current core functionality depends on those implementations and by overriding implementations without calling `super`, crazy things could happen.

To prevent this, (almost) identical protocol methods have been put in the `EventsProtocol`, which functions in parallel, but completely independent, of the native `Segnify` and `UIPageViewController` protocols.

***Relevant issues***
#8 

***Changes***
- What does it do?
It introduces similar or in some cases even identical protocol methods, in order to provide extended support for getting notified about several events. `EventsProtocol` forms the core of this functionality.

- In summary, what changes are made to the code?
`EventsProtocol` has been added.

***Check-list***
- [ ] Acceptance criteria in issue satisfied or Bug fixed
- [ ] Documentation present (relevant information is properly documented)
- [ ] Written tests for test suite (optional for the moment)
- [ ] Tested on devices
- [ ] CHANGELOG updated

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -189,6 +189,17 @@ Take a look at the [Segnified](https://github.com/nedap/Segnify/blob/master/Segn
...
+
+Implement `EventsProtocol` for the ability to be informed about several events, caused by either the `Segnify` instance or the `UIPageViewController` instance:
+
+- Getting the view controller before the currently displayed one
+- Getting the view controller after the currently displayed one
+- Gesture-initiated transition triggers
+- Segment selection trigger
+
+The protocol partly relies on the `UIPageViewControllerDataSource` and `UIPageViewControllerDelegate` protocols. All protocol methods are optional.
+
```
Comment (martias) [inline README.md]: nice 👍

File: Segnify/Protocols/EventsProtocol.swift [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    func segnify(_ segnify: Segnify, receivedTouchInside segment: Segment)
+}
+
+// MARK: - Defaults
+
+extension EventsProtocol {
+    
+    public func pageViewController(_ pageViewController: UIPageViewController,
+                                   requested viewController: UIViewController?,
+                                   before previousViewController: UIViewController) {}
```
Comment (martias) [inline Segnify/Protocols/EventsProtocol.swift]: haha, again the naming dilema :( . Not sure about the names `requested` , `after`, and `before` . I would like something more like `next`, `previous`, ...

File: Segnify/View Controllers/PageViewController.swift [hunk 3]
```diff
@@ -178,22 +186,30 @@ extension PageViewController: UIPageViewControllerDataSource {
...
         if previousIndex >= 0 {
             // Just return the previous view controller.
-            return contentElements[previousIndex].viewController
+            viewControllerToReturn = contentElements[previousIndex].viewController
         }
         else if segnify.delegate?.isScrollingInfinitely == true {
             // When `previousIndex` becomes negative, the user wants to scroll backwards from the first page.
             // Show the last page.
-            return contentElements.last!.viewController
+            viewControllerToReturn = contentElements.last!.viewController
```
Comment (martias) [inline Segnify/View Controllers/PageViewController.swift]: Would it be possible to do a safe unwrap here?. Swift lint rules from @dregatos are going to be sad if they find `!` 😢

File: Segnify/Views/Segnify.swift [hunk 4]
```diff
@@ -313,11 +316,12 @@ extension Segnify {
         // Set the current index.
         currentIndex = stackView.arrangedSubviews.firstIndex(of: selectedSegment!)!
         
-        // Notify the events delegate.
-        eventsDelegate?.didSelect(segment: selectedSegment!,
+        // Notify the events delegates.
+        segnifyEventsDelegate?.didSelect(segment: selectedSegment!,
```
Comment (martias) [inline Segnify/Views/Segnify.swift]: Same here, would it be possible to so safe unwrap ?

File: Segnify/View Controllers/PageViewController.swift (hunk 3 above)
Comment (barthopster) [reply Segnify/View Controllers/PageViewController.swift]: Fixed in https://github.com/nedap/Segnify/commit/13b31515dcbc729fb649cd6fc09647340bc56d34

File: Segnify/Views/Segnify.swift (hunk 4 above)
Comment (barthopster) [reply Segnify/Views/Segnify.swift]: Fixed in https://github.com/nedap/Segnify/commit/13b31515dcbc729fb649cd6fc09647340bc56d34

**Comment by barthopster** [comment]
Thanks for your feedback @martias.

> Are you solving any particular problem that you encountered while implementing Segnify in Dossier with this PR?
Yes. The implementation in the Dossier app is based on the `UIPageViewControllerDataSource` and `UIPageViewControllerDelegate` protocols. Segnify makes use of those protocols as well. Setting the `delegate` in Dossier would break Segnify and the other way around would break the functionality in Dossier, so I had to think of an alternative.
By 'copying' the protocol methods into `EventsProtocol`, the methods of the `UIPageViewControllerDataSource` and `UIPageViewControllerDelegate` protocols are getting 'forwarded', as you'll see in the code. In this way, both Segnify and its implementation into the Dossier app can use these protocols, without the need of overriding or exposing methods.

File: Segnify/View Controllers/PageViewController.swift [hunk 5]
```diff
@@ -29,8 +29,8 @@ open class PageViewController: UIViewController {
     /// A `Segnify` instance will be shown above the `PageViewController` instance, showing all `Segment` instances.
     public lazy var segnify: Segnify = {
         let segnify = Segnify()
-        segnify.eventsDelegate = self
         segnify.segnicator = Segnicator(configuration: DefaultSegnicatorDelegate())
+        segnify.segnifyEventsDelegate = self
```
[the remark this answers is not in the document]
Comment (dregatos) [reply Segnify/View Controllers/PageViewController.swift]: What about renaming it to `forwadedEventsDelegate`? And also, I think that this should NOT be `public`. The `SegnifyEventsProtocol` should be declared `internal` and hide it to the end user. My point is that the purpose of that protocol is to coordinate the Segnify and the page controller, right? That should be transparent for the end user.

[the remark this answers is not in the document]
Comment (barthopster) [reply Segnify/View Controllers/PageViewController.swift]: I agree with you, no need to expose `SegnifyEventsProtocol`. Fixed that, together with the renaming of `EventsProtocol` to `ForwardedEventsProtocol` in 
https://github.com/nedap/Segnify/commit/aa03673f94d85cc22f749051ee6d915026b343bb

**Comment by barthopster** [comment]
> I am a bit confused about all the different protocols that we have on Segnify.

I can imagine, as there are quite some protocols indeed. I aimed at a high level of customisation and compatibility, which is done the easiest by using protocols. 
If you know a different / easier / faster approach, please create an issue so we can improve it in the future.
Hopefully everything makes sense when you look at the implementation in Dossier 👍
